server/WFCORE-6750-unstable-api-annotation-scanner.adoc

```diff
@@ -0,0 +1,243 @@
+** The deployment unit processors will use the WildFly classloading rules to make sure that in the case of nested archives, that we only scan each class once. E.g. if a .war contains a `WEB-INF/lib/my-library.jar`, classes in `WEB-INF/lib/my-library.jar` will only be s scanned as part of the `WEB-INF/lib/my-library.jar` resource root scan, and not for the parent .war.
+** Once scanned, usage of annotated API elements may be reported depending on configuration in the `core-management` subsystem:
+*** A new child resource (currently called `service=unstable-api-annotations`) is added with a `level` attribute to configure the reporting
+**** Since this is a resource/feature with a stability level of preview, if the resource is not present, no reporting is done.
+****** The resource is added to the `preview` flavour of the current version of the model. At the time of writing it I was told there is no need to bump the version of management model constructs coming in lower than `community` level.
+**** If the attribute is set to `log` (this is the default value for the attribute), the usages will be logged to the server.log. In this case the user application will be deployed.
+**** If the attribute is set to `error`, the deployment will fail with an exception containing the usages.
+
+
+
+=== Nice-to-Have Requirements
+// Requirements in this section do not have to be met to merge the proposed functionality.
+// Note: Nice-to-have requirements that don't end up being implemented as part of
+// the work covered by this proposal should be moved to the 'Future Work' section.
+
+
+=== Non-Requirements
+// Use this section to explicitly discuss things that readers might think are required
+// but which are not required.
+* Full indexing of all annotations indicating unstable API
+** See the 'Future Work' section for plans to get full coverage
+* Inspection of user classes not in an application deployment, such as:
+** User provided modules
+** Classes provided by the user via the management model with the typical `module`/`class` attribute pairs.
+* No attempt is made to not add the indices if the server is provisioned a a higher stability level than 'preview'. In other words, if the feature packs contain the indices, they will always end up in the server.
+* Since this is a `preview` feature, and the model version is the same as the original with the added `preview` qualifier, model transformation is not needed.
+
+=== Future Work
+// Use this section to discuss requirements that are not addressed by this proposal
+// but which may be addressed in later proposals.
+* Promote this feature to Community or higher.
+** This would allow us to enable it out of the box by default, making it more useful.
+* Identify all annotations indicating unstable API used by WildFly itself with the help of component leads.
+** Scan for these annotations during feature pack builds, and make them available in a provisioned server.
```

Comment:
What would be the goal of this?

Reply:
To index similar annotations from other libraries than the current SmallRye and Hibernate families. E.g. I think JGroups has something. Apache might have something etc. i.e. get better coverage

Reply:
My question here is more about why we make them available in the provisioned server. But probably I misinterpreted the parent point.

If your meaning in the parent point is just to include more annotations from other components and check for their use in deployment code, I understand. If you're talking about scanning WF code for our own use of things, I understand why we'd do that, I'm just not clear why we'd want to write the result in the provisioned server.

In any case I'm not asking for any change in the document; this is just a discussion of what you had in mind.